### ISSUE TITLE
MMOD loss may go negative if detection doesn't match a truth box but does match an ignore box

### ISSUE THREAD

**Issue by reunanen** [davisking/dlib#1894]
## Expected Behavior
MMOD loss never goes negative for valid input.

(Note that what constitutes "valid input" may need to be discussed. In my case, there are correct truth-rects that are very close to each other. They also overlap so much that it does not look like a great idea to try to teach all of them. So I'm setting a stricter `overlaps_nms`, and setting some of the truth-rects to be ignored. However, I don't want to completely disregard any of them (they contain valuable information), so for different epochs I'm randomly selecting which rects are to be ignored. Thought this should be fine, as long as there are no non-ignored rects that overlap, as per the `overlaps_nms` that I have set.)

## Current Behavior
When there are ignore-rects very close to and overlapping with the (non-ignore) truth-rects, it looks like it's possible that `truth_match <= options.truth_match_iou_threshold`, while `overlaps_ignore_box(*truth, dets[i].rect) == true`. In other words, a candidate detection doesn't quite match the (correctly-)best-matching non-ignore rect with `> options.truth_match_iou_threshold`, but does overlap an ignore-box.

As a result, it seems that the loss may go negative. (Specifically because in this case the candidate detection isn't added to `final_dets`, and therefore the corresponding `out_data` isn't added to `loss`.)

Note that it's not just the loss that goes sour; also the training itself does not really "work", i.e. the gradients are not updated correctly it seems. (This seems to hold at least in a real-world project I'm working with, where the network after a long training run eventually detects nothing – whereas a shorter training run works ok. But for this gradient part I don't have how-to-reproduce instructions, at least just yet. However, I guess loss and gradient go hand-to-hand, so possibly fixing the negative loss will fix the gradient updates as well.)

## Workarounds
I have tried the following workarounds:
- Set `options.truth_match_iou_threshold = 0.25` (arbitrary selection).
- Change code so that the gradient is not updated for samples for which the contribution to loss would be negative: https://github.com/reunanen/dlib/commit/0d55d3cf1bd9cd082f5bd8361fe551972c92ee61

While these approaches seem to alleviate the problem to some degree at least, I can't say I'm very happy with either of them. (Think they are only band-aid at best.)

## Steps to Reproduce

Clone current `master` branch, and cherry-pick this: https://github.com/reunanen/dlib/commit/781065f62f57ccc4466aa5e110f66642e99bc438

Run the dnn tests – the newly added test should fail.

**Comment by reunanen** [comment]
So while ideally I would like the code to Just Work (tm), also suggestions how to deal with these kinds of situations at the callsite are very welcome. (If parameter tuning is the way to go, then suggestions how to do it programmatically depending on the input truth-boxes would be appreciated.)

**Comment by davisking** [comment]
The loss should only end up going negative if it's impossible to produce the requested label output, given the settings.  There is some auto-ignoring already in the code when a truth box is obviously outside the area the cnn can find.  If the `image_rect_to_feat_coord()` method maps a truth box to a point covered by an ignore box maybe it should auto-ignore that truth box too.  That sounds like the root of your problem.  I'm not sure this is such a great idea though.  It's kinda funny to have truth boxes that overlap ignore regions.  

Really you should train with the labels you want to detect.  Why can't you do that in your case? Just because the NMS rules it out?  One way or another you have to train with a set of labels that's achievable by the detector.

**Comment by reunanen** [comment]
Thanks for your prompt reply, Davis.

> The loss should only end up going negative if it's impossible to produce the requested label output, given the settings.

Yep, have generally understood this for a while already. But guessing had missed that this apparently applies to ignore-rects as well. (Didn't really think ignore-rects were part of "the requested label output".)

> If the image_rect_to_feat_coord() method maps a truth box to a point covered by an ignore box maybe it should auto-ignore that truth box too. That sounds like the root of your problem. I'm not sure this is such a great idea though.

Yeah – doesn't sound great to me either. I don't really want to have the truth-boxes ignored in this situation. I do want a detection for the non-ignored truth box. I just don't want to make the network learn that there's nothing in the overlapping rect (because I have to assume that there is). Instead, I wanted the other rect to have no effect whatsoever (in terms of gradient or loss).

> Really you should train with the labels you want to detect. Why can't you do that in your case? Just because the NMS rules it out? One way or another you have to train with a set of labels that's achievable by the detector.

Yeah, I guess I need to take a step back maybe. At some point, I was receiving annotations that were overlapping a lot. (When I say "I", I mean an automated system that I've been writing – so I'm not able to tweak parameters manually, and I can't expect that the users will do it either. So the default behaviour should be at least somehow sane, even if the performance can be improved by problem-specific parameter tuning).

I am sure some of these highly overlapping annotations were user errors. On the other hand, some of them appeared to be perfectly valid; but I thought they might be too challenging (if not next to impossible) to learn correctly.

(Think of something like [this](https://serc.carleton.edu/images/advancegeo/resources/overlapping_hands.v2_650.jpg), for example. Suppose we have a hand detector, which is generally trained by drawing rects tightly around just hands (not arms). Then for the example image it appears that it's not exactly wrong to draw eight almost completely overlapping rects (well all of them might not exactly overlap... but for the sake of argument, let's pretend they do). Now, for a human being it's pretty clear from the context that there are in total 8 hands there, but I'm afraid it might be a bit too much for a neural network to figure out, if there's not extensive training material available. Or what do you think? This picture of hands is a toy example of course, but I think the challenge is still real: objects can overlap a lot in images, so at least such annotations are not incorrect, right? (Regardless of whether it's possible to detect all these objects separately.) [Here's another example.](https://www.wonderopolis.org/_img?img=/wp-content/uploads/video_thumbs/10654563.jpg) While I'm certainly not an expert in the sheep domain, I think a complete set of annotations would overlap a whole lot. Note that I'm not saying I need to detect every individual sheep (although that would be wonderful); instead, I want to have a training process that behaves _reasonably_, when presented this kind of annotation input. Either with a fixed set of default parameters, or some parameters I can figure out programmatically (given the truth-rects).)

So what I did previously was set NMS to something like 70% by default, although the incoming data would have suggested a much higher value, e.g. 95%. Limiting the NMS threshold by itself [created problems](https://github.com/davisking/dlib/issues/870#issuecomment-336427007), so I started to auto-ignore one from each pair (or two from each triplet, or three from each quadruplet... but alternating from one epoch to another). (Here, by auto-ignoring I mean I set the ignore flag on the application side of it, and not inside dlib.)

But looking at this again now, it seems that it's not enough to simply set those truth rects (which overlap more than what is allowed by `overlaps_nms`) to be ignored.

**Comment by reunanen** [comment]
I will look more into this, and hopefully come up with some solution / suggestion.
